src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -0,0 +1,551 @@
+    )
+    optional.add_argument(
+        "--keep_job_dir",
+        action="store_true",
+        help="Keep SDK job directory after upload task is completed",
+    )
+    return parser
+
+
+def _get_yaml_file_path(root_dir: str, assembly_dir_name: str) -> str:
+    """
+    Get the uploaded.yaml file path from NCBI/sourcedata directory.
+    """
+    file_path = os.path.join(get_work_dir(root_dir), assembly_dir_name, UPLOADED_YAML)
+    Path(file_path).touch(exist_ok=True)
+    return file_path
+
+
+def _sanitize_data_dir(job_dir: str) -> str:
+    """
+    Create a temporary directory for storing uploaded files.
+    """
+    data_dir = os.path.join(job_dir, "workdir/tmp", DATA_DIR)
+    if os.path.exists(data_dir):
+        shutil.rmtree(data_dir)
+    os.makedirs(data_dir)
+    return data_dir
+
+
+def _get_source_file(assembly_dir: str, assembly_file: str) -> str:
+    """
+    Get the sourcedata file path from the assembly directory.
+    """
+    if not os.path.islink(assembly_dir):
+        raise ValueError(f"{assembly_dir} is not a symlink")
+    src_file = os.path.join(os.readlink(assembly_dir), assembly_file)
+    return src_file
+
+
+def _upload_assembly_to_workspace(
+    conf: Conf,
+    workspace_name: str,
+    file_path: str,
+    assembly_name: str,
+) -> None:
+    """Upload an assembly file to workspace."""
+    success, attempts, max_attempts = False, 0, 3
+    while attempts < max_attempts and not success:
+        try:
+            time.sleep(attempts)
+            conf.asu.save_assembly_from_fasta(
+                {
+                    "file": {"path": file_path},
+                    "workspace_name": workspace_name,
+                    "assembly_name": assembly_name,
+                }
+            )
+            success = True
+        except Exception as e:
+            print(f"Error:\n{e}\nfrom attempt {attempts + 1}.\nTrying to rerun.")
+            attempts += 1
+
+    if not success:
+        raise ValueError(
+            f"Upload Failed for {file_path} after {max_attempts} attempts!"
+        )
+
+
+def _read_yaml_file(
+        root_dir: str,
+        env: str,
+        workspace_id: int,
+        assembly_dir_name: str,
+        assembly_name: str, 
+) -> Tuple[dict[str, dict[int, list[str]]], bool]:
+    """
+    Get metadata and upload status of an assembly from the uploaded.yaml file.
+    """
+
+    uploaded = False
+    if env not in loader_common_names.KB_ENV:
+        raise ValueError(f"Currently only support these {loader_common_names.KB_ENV} envs for upload")
+
+    file_path = _get_yaml_file_path(root_dir, assembly_dir_name)
+
+    with open(file_path, "r") as file:
+        data = yaml.safe_load(file)
+
+    if not data:
+        data = {env: dict()}
+
+    if workspace_id not in data[env]:
+        data[env][workspace_id] = list()
+
+    assembly_dicts = data[env][workspace_id]
+    if assembly_name in [assembly_dict["file_name"] for assembly_dict in assembly_dicts]:
+        uploaded = True
```

Comment:
This assumes there could be > 1 assembly file to upload per genome dir. is that the case? I thought not

Reply:
So, it might be uploaded to diff env or workspace_id, but with the same env/workspace_id, assembly file should be 1. Yeah, thank you for correcting me!

Reply:
Updated!

Reply:
:+1: